packages/backend-storage/src/storage_access_policy_factory.ts

```diff
@@ -69,20 +70,48 @@ export class StorageAccessPolicyFactory {
 
   private getStatement = (
     s3Prefixes: Readonly<Set<StoragePath>>,
-    action: StorageAction,
+    action: InternalStorageAction,
     effect: Effect
-  ) =>
-    new PolicyStatement({
-      effect,
-      actions: actionMap[action],
-      resources: Array.from(s3Prefixes).map(
-        (s3Prefix) => `${this.bucket.bucketArn}${s3Prefix}`
-      ),
-    });
+  ) => {
+    switch (action) {
+      case 'delete':
+      case 'get':
+      case 'write':
+        return new PolicyStatement({
+          effect,
+          actions: actionMap[action],
+          resources: Array.from(s3Prefixes).map(
+            (s3Prefix) => `${this.bucket.bucketArn}${s3Prefix}`
+          ),
+        });
+      case 'list':
+        return new PolicyStatement({
+          effect,
+          actions: actionMap[action],
+          resources: [this.bucket.bucketArn],
+          conditions: {
+            StringLike: {
+              's3:prefix': Array.from(s3Prefixes).flatMap(toConditionPrefix),
+            },
+          },
+        });
+    }
+  };
 }
 
-const actionMap: Record<StorageAction, string[]> = {
-  read: ['s3:GetObject'],
+const actionMap: Record<InternalStorageAction, string[]> = {
+  get: ['s3:GetObject'],
+  list: ['s3:ListBucket'],
   write: ['s3:PutObject'],
   delete: ['s3:DeleteObject'],
 };
+
+/**
+ * Converts a prefix like /foo/bar/* into [foo/bar/, foo/bar/*]
+ * This is necessary to grant the ability to list all objects directly in "foo/bar" and all objects under "foo/bar"
+ */
```

Comment:
If we have to remove the `/` prefix for the users, why don't we just have the user not use `/` at the beginning of the path? also, using `/` at the beginning means the root dir, but without `/` means the cwd, I think we mean cwd in our case, right?

Reply:
These are s3 prefixes, not local paths. The `/` at the beginning is meant to make it clear that all paths should start from the bucket root.